### PR TITLE
Feature newtonsolver

### PR DESCRIPTION
**Pull request by paulstapor**
To be merged in some near future, currently only for CI tests.
Current aim: get most important linear solvers to work. Already done:
 - BiCGStab (hacked version by Glenn)
 - cvDense (modified kind of calling the solver compared to ODE integration, since only little parts are needed)

To be done soon:
 - KLU (sparse solver)
 - GMRES (iterative solver)

### CODE REVIEW

File: @amioption/amioption.m [hunk 1]
```diff
@@ -248,37 +245,30 @@
...
         end
 
         function this = set.newton_preeq(this,value)
-            assert(isnumeric(value),'The option preequil must have a numeric value!')
-            assert(floor(value)==value,'The option preequil must be an integer!')
+            assert(isnumeric(value),'The option preequil must have a logical value!')
+            assert(floor(value)==value,'The option preequil must be a logical!')
             assert(value<=2,'Only 0 and 1 are valid options for preequil!')
-            this.preequil = value;
+            this.newton_preequil = value;
```
Comment (dweindl) [inline @amioption/amioption.m]: newton_preeq vs newton_preequil, assignment to wrong variable?!

File: auxiliary/compileAMICIDependencies.m [hunk 2]
```diff
@@ -71,7 +71,7 @@
 function includesstr = getIncludeString(wrap_path, sundials_path, ssparse_path)
     includesstr = '';
     includesstr = strcat(includesstr,' -I"', fullfile(sundials_path, 'include'), '"');
-    includesstr = strcat(includesstr,' -I"', fullfile(sundials_path, 'src','cvodes'), '"');
+    includesstr = strcat(includesstr,' -I"', fullfile(sundials_path, 'src'), '"');
```
Comment (dweindl) [inline auxiliary/compileAMICIDependencies.m]: Can remove this line again, this was only for the `*_impl.h`

File: include/newton_solver.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+
+
+class NewtonSolverSparse : public NewtonSolver {
+    
+public:
+    NewtonSolverSparse(Model *model, ReturnData *rdata, UserData *udata, TempData *tdata);
+    int getStep(int ntry, int nnewt, N_Vector delta);
+    ~NewtonSolverSparse();
+    klu_common common;
```
Comment (dweindl) [inline include/newton_solver.h]: Can't this be private?

File: src/CMakeLists.template.txt [hunk 4]
```diff
@@ -25,6 +25,7 @@ include_directories("${AMICI_DIR}")
 include_directories("${CMAKE_CURRENT_SOURCE_DIR}")
 include_directories("${HDF5_INCLUDE_DIRS}")
 include_directories("${AMICI_DIR}/sundials/include")
+include_directories("${AMICI_DIR}/sundials/src")
```
Comment (dweindl) [inline src/CMakeLists.template.txt]: This line can be removed, see above

File: src/amici_interface_matlab.cpp [hunk 5]
```diff
@@ -169,6 +169,8 @@ UserData *userDataFromMatlabCall(const mxArray *prhs[], int nrhs, Model *model)
         readOptionScalar(ism,int)
         readOptionScalar(sensi_meth,AMICI_sensi_meth)
         readOptionScalar(ordering,int)
+        readOptionScalar(newton_preeq,int)
```
Comment (dweindl) [inline src/amici_interface_matlab.cpp]: Check name, see above.

File: src/newton_solver.cpp [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+     *
+     * @param[in] udata pointer to the user data struct @type UserData
+     * @param[out] rdata pointer to the return data struct @type ReturnData
+     * @param[in] tdata pointer to the temporary data struct @type TempData
+     * @param[out] tdata pointer to the temporary data struct @type TempData
+     * @param[in] ami_mem pointer to the solver memory block @type *void
+     * @return int status flag indicating success of execution @type int
+     */
+    
+    NewtonSolver *newtonSolver;
```
Comment (dweindl) [inline src/newton_solver.cpp]: Can remove this line.

File: src/steadystateproblem.cpp [hunk 7]
```diff
@@ -206,21 +214,121 @@ int SteadystateProblem::applyNewtonsMethod(UserData *udata, ReturnData *rdata, T
...
+int SteadystateProblem::linsolveSPBCG(UserData *udata, ReturnData *rdata, TempData *tdata, int ntry, int nnewt, N_Vector ns_delta, Model *model) {
+    /**
+         * linsolveSPBCG solves the linear system for the Newton iteration by using the BiCGStab algorithm.
+         * This routines is to be stored in another file in near future.
          *
          * @param[in] udata pointer to the user data struct @type UserData
          * @param[out] rdata pointer to the return data struct @type ReturnData
          * @param[in] tdata pointer to the temporary data struct @type TempData
          * @param[out] tdata pointer to the temporary data struct @type TempData
+         * @param[in] ami_mem pointer to the solver memory block @type *void
```
Comment (dweindl) [inline src/steadystateproblem.cpp]: obsolete

File: src/udata.cpp [hunk 8]
```diff
@@ -60,6 +60,8 @@ void UserData::init()
     maxsteps = 0;
     newton_maxsteps = 0;
     newton_maxlinsteps = 0;
+    newton_precon = 1;
+    newton_preeq = 0;
```
Comment (dweindl) [inline src/udata.cpp]: If there is only 0/1 please use TRUE/FALSE, otherwise ok

Comment (paulstapor) [inline src/udata.cpp]: Done

File: src/amici_interface_matlab.cpp (hunk 5 above)
Comment (paulstapor) [inline src/amici_interface_matlab.cpp]: Done

File: @amioption/amioption.m (hunk 1 above)
Comment (paulstapor) [inline @amioption/amioption.m]: Yes, sorry.

File: include/newton_solver.h (hunk 3 above)
Comment (paulstapor) [inline include/newton_solver.h]: Yes, it can. And now, it is...

File: auxiliary/compileAMICIDependencies.m (hunk 2 above)
Comment (paulstapor) [inline auxiliary/compileAMICIDependencies.m]: Done.

File: src/CMakeLists.template.txt (hunk 4 above)
Comment (paulstapor) [inline src/CMakeLists.template.txt]: Done.

File: src/newton_solver.cpp (hunk 6 above)
Comment (paulstapor) [inline src/newton_solver.cpp]: Removed, as discussed.

File: src/steadystateproblem.cpp (hunk 7 above)
Comment (paulstapor) [inline src/steadystateproblem.cpp]: changed

**Pull request merged**